daemon/cdi_linux.go

```diff
@@ -0,0 +1,85 @@
+package daemon
+
+import (
+	"fmt"
+
+	"github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	"github.com/docker/docker/pkg/capabilities"
+	"github.com/hashicorp/go-multierror"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+)
+
+// cdiDeviceInjector defines the operations required to inject CDI devices into an OCI spec.
+type cdiDeviceInjector interface {
+	InjectDevices(*specs.Spec, ...string) ([]string, error)
+	GetErrors() map[string][]error
+}
+
+type cdiHandler struct {
+	registry cdiDeviceInjector
+}
+
+// RegisterCDIDriver registers the CDI device driver.
+// The driver injects CDI devices into an incoming OCI spec and is called for DeviceRequests associated with CDI devices.
+func RegisterCDIDriver(opts ...cdi.Option) {
+	var registry cdiDeviceInjector
+
+	cache, err := cdi.NewCache(opts...)
+	if err != nil {
+		// We store an invalid cache that always returns an error.
+		// This error will be returned when a CDI device is requested. This ensures that daemon startup is not blocked
+		// by a CDI registry initialization failure.
+		registry = &invalidCache{err: fmt.Errorf("failed to initialize CDI registry: %w", err)}
+	} else {
+		registry = cache
+	}
+	c := &cdiHandler{
+		registry: registry,
+	}
+
+	driver := &deviceDriver{
+		capset:     capabilities.Set{"cdi": struct{}{}},
+		updateSpec: c.injectCDIDevices,
+	}
+
+	registerDeviceDriver("cdi", driver)
+}
+
+// injectCDIDevices injects a set of CDI devices into the specified OCI specification.
+func (c *cdiHandler) injectCDIDevices(s *specs.Spec, dev *deviceInstance) error {
+	if dev == nil {
+		return nil
+	}
+
+	cdiDeviceNames := dev.req.DeviceIDs
```

Comment:
Is there a suggested error type to use here? (`InvalidParameter`)?

Reply:
The Nvidia driver uses [a unique and otherwise unreferenced error](https://github.com/moby/moby/blob/17d94cf3b901f52a70c268c8971b8887baa215e4/daemon/nvidia_linux.go#L18) in a similar case, but I think that's probably a bit legacy, and I would certainly _suggest_ that using `errdefs.InvalidParameter` is the right way to go. (The nvidia driver should probably be made the same, as part of some future refactoring, I guess; I suspect right now the similar nvidia driver error [turns into a 500 error in the API](https://github.com/moby/moby/blob/17d94cf3b901f52a70c268c8971b8887baa215e4/api/server/httpstatus/status.go#L74-L76), instead of the [400 you'd expect](https://github.com/moby/moby/blob/17d94cf3b901f52a70c268c8971b8887baa215e4/api/server/httpstatus/status.go#L35-L36)).

The [error raised when no matching device plugin is found](https://github.com/moby/moby/blob/17d94cf3b901f52a70c268c8971b8887baa215e4/daemon/errors.go#L83-L92) also resolves into `errdefs.InvalidParameter` (and hence returns 400 in the API).